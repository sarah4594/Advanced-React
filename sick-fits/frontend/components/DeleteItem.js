/* eslint-disable no-unused-vars */
import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import PropTypes from 'prop-types'
import gql from 'graphql-tag'
import { ALL_ITEMS_QUERY } from './Items'

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`

class DeleteItem extends Component {
  update = (cache, payload) => {
    // manually update cache on client so it matches server
    // 1. Read the cache for the item we want (which query
    // did we use to get items onto the page?)
    // (cannot reach directly into cache to delte
    // one item so we have to query to read items
    // from cache and write back into cache)
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY })
    console.log(data, payload)
    // 2. Filter deleted item out of page
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id,
    )
    // 3. Put items back
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data })
  }
  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(deleteItem, { error }) => (
          <button
            onClick={() => {
              if (confirm('Are you sure you want to delete this item?')) {
                deleteItem().catch(err => {
                  alert(err.message)
                })
              }
            }}
          >
            {this.props.children}
          </button>
        )}
      </Mutation>
    )
  }
}

DeleteItem.propTypes = {
  id: PropTypes.string.isRequired,
  children: PropTypes.string,
}

export default DeleteItem
