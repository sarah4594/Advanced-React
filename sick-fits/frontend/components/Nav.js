import SignedInNav from './NavBar/SignedInNav'
import SignedOutNav from './NavBar/SignedOutNav'
import User from './User'

const Nav = () => (
  <User>
    {({ data }) => {
      const me = data ? data.me : null
      if (!me) return <SignedOutNav />
      return <SignedInNav me={me} />
    }}
  </User>
)

export default Nav
