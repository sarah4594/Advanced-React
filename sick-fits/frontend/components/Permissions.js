/* eslint-disable no-unused-vars */
import React from 'react'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import PropTypes from 'prop-types'
import Error from './ErrorMessage'
import Table from './styles/Table'
import SickButton from './styles/SickButton'

const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
]

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    users {
      id
      name
      email
      permissions
    }
  }
`

const Permissions = props => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) =>
      console.log(data) || (
        <div>
          <Error error={error} />
          <div>
            <h2>Manage Permissions</h2>
            <Table>
              <thead>
                <tr>
                  <th>Name</th>
                  <th>Email</th>
                  {possiblePermissions.map(permission => (
                    <th key={permission}>{permission}</th>
                  ))}
                  <th>Update</th>
                </tr>
              </thead>
              <tbody>
                {data.users.map(user => (
                  <UserPermissions user={user} key={user.id} />
                ))}
              </tbody>
            </Table>
          </div>
        </div>
      )
    }
  </Query>
)

class UserPermissions extends React.Component {
  // Usually do not want to use props to set initial state
  // Exception: seeding - the initial data is coming from
  // props but whenever someone changes a permission inside
  // of the user then it is going to be updated to state
  // then when you click save it will be sent of to backend
  state = {
    permissions: this.props.user.permissions,
  }

  handlePermissionChange = e => {
    const checkbox = e.target
    // take a copy of current permissions
    let updatedPermissions = [...this.state.permissions]
    // Figure put if we need to add permission...
    if (checkbox.checked) {
      // Add it in
      updatedPermissions.push(checkbox.value)
      // ...Or remove permission
    } else {
      updatedPermissions = updatedPermissions.filter(
        permission => permission !== checkbox.value,
      )
    }
    this.setState({ permissions: updatedPermissions })
  }

  render() {
    const user = this.props.user
    return (
      <tr>
        <td>{user.name}</td>
        <td>{user.email}</td>
        {possiblePermissions.map(permission => (
          <td key={permission}>
            <label htmlFor={`${user.id}-permission-${permission}`}>
              <input
                type="checkbox"
                checked={this.state.permissions.includes(permission)}
                value={permission}
                onChange={this.handlePermissionChange}
              />
            </label>
          </td>
        ))}
        <td>
          <SickButton>Update</SickButton>
        </td>
      </tr>
    )
  }
}

UserPermissions.propTypes = {
  user: PropTypes.shape({
    name: PropTypes.string,
    email: PropTypes.string,
    id: PropTypes.string,
    permissions: PropTypes.array,
  }).isRequired,
}

export default Permissions
