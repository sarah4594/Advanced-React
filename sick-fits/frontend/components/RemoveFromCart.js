import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import gql from 'graphql-tag'
import { CURRENT_USER_QUERY } from './User'
import { Mutation } from 'react-apollo'

const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`

class RemoveFromCart extends React.Component {
  // this gets called as soon as we get a response back from the server
  // after a mutation has been performed
  update = (cache, payload) => {
    console.log('running')
    // 1. Read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY })
    console.log(data)
    // 2. Remove item form cart
    const cartItemId = payload.data.removeFromCart.id
    data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId)
    // 3. Write it back to the cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data })
  }
  render() {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(removeFromCart, { loading, error }) => (
          <BigButton
            disabled={loading}
            onClick={() => {
              removeFromCart().catch(err => alert(err.message))
            }}
            title="Delete Item"
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    )
  }
}

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired,
}

export default RemoveFromCart
