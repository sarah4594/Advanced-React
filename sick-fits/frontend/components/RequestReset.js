import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Form from './styles/Form'
import Error from './ErrorMessage'

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`

class RequestReset extends Component {
  state = {
    name: '',
  }

  saveToState = e => {
    this.setState({ [e.target.name]: e.target.value })
  }

  render() {
    return (
      <Mutation mutation={REQUEST_RESET_MUTATION} variables={this.state}>
        {/* called = boolean that says whether or not mutation
        has been called yet */}
        {(reset, { error, loading, called }) => {
          return (
            <Form
              method="post"
              onSubmit={async e => {
                e.preventDefault()
                await reset()
                this.setState({ email: '' })
              }}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Forgot Your Password</h2>
                <Error error={error} />
                {!error && !loading && called && (
                  <p>Success! Check you email for a reset link</p>
                )}
                <label htmlFor="email">
                  Email
                  <input
                    type="email"
                    name="email"
                    placeholder="email"
                    value={this.state.email}
                    onChange={this.saveToState}
                  />
                </label>

                <button type="submit">Reset Password</button>
              </fieldset>
            </Form>
          )
        }}
      </Mutation>
    )
  }
}

export default RequestReset
