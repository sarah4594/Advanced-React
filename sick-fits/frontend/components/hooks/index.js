import { useState } from 'react'

function useFormInput(initialValue) {
  const [value, setValue] = useState(initialValue)
  function handleChange(e) {
    setValue(e.target.value)
  }
  return {
    value,
    onChange: handleChange,
  }
}

function useCurrentInput(initialValue) {
  const [value, setValue] = useState(initialValue)
  function handleChange(e) {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    setValue({ [name]: val })
  }
  return {
    value,
    onChange: handleChange,
  }
}

export { useFormInput }
export { useCurrentInput }
