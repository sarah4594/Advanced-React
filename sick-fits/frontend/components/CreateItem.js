import React, { useState } from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Router from 'next/router'
import Form from './styles/Form'
import Error from './ErrorMessage'
import { useFormInput } from './hooks'

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

function CreateItem() {
  // state = {
  //   title: '',
  //   description: '',
  //   image: '',
  //   largeImage: '',
  //   price: 0,
  // }
  // handleChange = e => {
  //   const { name, type, value } = e.target
  //   const val = type === 'number' ? parseFloat(value) : value
  //   this.setState({ [name]: val })
  // }

  const title = useFormInput('')
  const description = useFormInput('')
  const [image, setImage] = useState('')
  const [largeImage, setLargeImage] = useState('')
  const price = useFormInput(0)

  async function uploadFile(e) {
    //console.log('Uplaoding file')
    // pull files out of selection
    const files = e.target.files
    // form data api (part of Javascript) to prep data
    const data = new FormData()
    data.append('file', files[0])
    // add upload preset
    data.append('upload_preset', 'sickfits')

    //hit cloudinary api
    const res = await fetch(
      'https://api.cloudinary.com/v1_1/snickersarah/image/upload',
      {
        method: 'POST',
        body: data,
      },
    )
    const file = await res.json()
    // parse data that comes back
    //console.log(file)
    // put the data into the spots they go into
    // this.setState({
    //   image: file.secure_url,
    //   largeImage: file.eager[0].secure_url,
    // })
    setImage(file.secure_url)
    setLargeImage(file.eager[0].secure_url)
    debugger
  }

  const handleSubmit = async (e, createItem) => {
    // stop page from submitting
    e.preventDefault()
    // call the mutation
    const res = await createItem()
    // change them to the single item page
    //console.log(res)
    Router.push({
      pathname: '/item',
      query: { id: res.data.createItem.id },
    })
  }
  return (
    <Mutation
      mutation={CREATE_ITEM_MUTATION}
      variables={{
        title: title.value,
        description: description.value,
        price: price.value,
      }}
    >
      {(createItem, { loading, error }) => (
        <Form data-test="form" onSubmit={e => handleSubmit(e, createItem)}>
          <Error error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
            <label htmlFor="file">
              Image
              <input
                type="file"
                id="file"
                name="file"
                placeholder="Upload an image"
                required
                onChange={uploadFile}
              />
              {image && <img width="200" src={image} alt="Upload Preview" />}
            </label>

            <label htmlFor="title">
              Title
              <input
                type="text"
                id="title"
                name="title"
                placeholder="Title"
                required
                {...title}
              />
            </label>

            <label htmlFor="price">
              Price
              <input
                type="number"
                id="price"
                name="price"
                placeholder="Price"
                required
                {...price}
              />
            </label>

            <label htmlFor="price">
              Description
              <textarea
                id="description"
                name="description"
                placeholder="Description"
                required
                {...description}
              />
            </label>

            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      )}
    </Mutation>
  )
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
