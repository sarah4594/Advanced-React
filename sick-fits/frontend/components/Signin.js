import React from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Router from 'next/router'
import Form from './styles/Form'
import Error from './ErrorMessage'
import { CURRENT_USER_QUERY } from './User'
import { useFormInput } from './hooks'

const SIGNIN_MUTATION = gql`
  mutation SIGNIN_MUTATION($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      id
      email
      name
    }
  }
`

function Signin() {
  // state = {
  //   name: '',
  //   password: '',
  //   email: '',
  // }
  // saveToState = e => {
  //   this.setState({ [e.target.name]: e.target.value })
  // }

  const name = useFormInput('')
  const password = useFormInput('')
  const email = useFormInput('')

  async function handleSubmit(e, signin) {
    e.preventDefault()
    const user = await signin()
    if (user) {
      Router.push({
        pathname: '/items',
      })
    }
  }

  return (
    <Mutation
      mutation={SIGNIN_MUTATION}
      variables={{
        name: name.value,
        password: password.value,
        email: email.value,
      }}
      refetchQueries={[{ query: CURRENT_USER_QUERY }]}
    >
      {(signin, { error, loading }) => {
        return (
          <Form method="post" onSubmit={e => handleSubmit(e, signin)}>
            <fieldset disabled={loading} aria-busy={loading}>
              <h2>Sign Into Your Account</h2>
              <Error error={error} />
              <label htmlFor="email">
                Email
                <input
                  type="email"
                  name="email"
                  placeholder="email"
                  {...email}
                />
              </label>
              <label htmlFor="password">
                Password
                <input
                  type="password"
                  name="password"
                  placeholder="password"
                  {...password}
                />
              </label>

              <button type="submit">Sign In!</button>
            </fieldset>
          </Form>
        )
      }}
    </Mutation>
  )
}

export default Signin
export { SIGNIN_MUTATION }
