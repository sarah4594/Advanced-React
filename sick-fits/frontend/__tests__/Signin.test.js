import { mount } from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import { ApolloConsumer } from 'react-apollo'
import { MockedProvider } from 'react-apollo/test-utils'
import Router from 'next/router'
import Signin, { SIGNIN_MUTATION } from '../components/Signin'
import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser } from '../lib/testUtils'

const me = fakeUser()

Router.router = {
  push() {},
}

function type(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: { name, value },
  })
}

const mocks = [
  // Current user query mock
  {
    request: {
      query: CURRENT_USER_QUERY,
    },
    result: { data: { me } },
  },
]

describe('<Signin />', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <Signin />
      </MockedProvider>,
    )
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot()
  })
})
