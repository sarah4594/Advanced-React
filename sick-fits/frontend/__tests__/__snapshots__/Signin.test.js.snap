// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signin /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 fCpNnG"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Into Your Account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign In!
    </button>
  </fieldset>
</form>
`;
